src/optimization/ControlFlow.cpp

```diff
@@ -1181,15 +1182,170 @@ void optimizations::addStartStopSegment(const Module& module, Method& method, co
+                    {
+                        edge.data.isBackEdge = true;
+                    }
+                    else
+                    {
+                        que.push(&next);
+                    }
+                    return true;
+                });
+            }
+        }
+
+        std::vector<CFGNode*> unnecessaryNodes;
+
+        for(auto& pair : cfg->getNodes())
+        {
+            // a.
+            bool hasConstantInstruction = false;
+            auto block = pair.first;
+            for(auto it = block->walk(); it != block->walkEnd(); it = it.nextInBlock())
+            {
+                if(it.has() && it->isConstantInstruction())
+                {
+                    hasConstantInstruction = true;
+                    break;
+                }
+            }
+            if(hasConstantInstruction)
+            {
+                continue;
+            }
+
+            int incomingEdgesCount = 0;
+            pair.second.forAllIncomingEdges([&incomingEdgesCount](const CFGNode& next, const CFGEdge& edge) -> bool {
+                incomingEdgesCount++;
+                return true;
+            });
+            int outgoingEdgesCount = 0;
+            pair.second.forAllOutgoingEdges([&outgoingEdgesCount](const CFGNode& next, const CFGEdge& edge) -> bool {
+                outgoingEdgesCount++;
+                return true;
+            });
+
+            // b.
+            if(incomingEdgesCount >= 2 || outgoingEdgesCount >= 2)
+            {
+                continue;
+            }
+
+            // c.
+            bool hasBackEdge = false;
+            pair.second.forAllOutgoingEdges([&hasBackEdge](const CFGNode& next, const CFGEdge& edge) -> bool {
+                next.forAllIncomingEdges([&hasBackEdge](const CFGNode& nnext, const CFGEdge& nedge) -> bool {
+                    if(nedge.data.isBackEdge)
+                    {
+                        hasBackEdge = true;
+                        return false;
+                    }
+                    return true;
+                });
+                if(hasBackEdge)
+                {
+                    return false;
+                }
+                return true;
+            });
+            if(hasBackEdge)
+            {
+                continue;
+            }
+
+            unnecessaryNodes.push_back(&pair.second);
+        }
+
+        for(auto& node : unnecessaryNodes)
+        {
+            CFGNode *prev = nullptr, *next = nullptr;
+            // A unnecessary node must has just one previous node and one next node.
+            node->forAllIncomingEdges([&prev](CFGNode& node, CFGEdge&) -> bool {
+                prev = &node;
+                return false;
+            });
+            node->forAllOutgoingEdges([&next](CFGNode& node, CFGEdge&) -> bool {
+                next = &node;
+                return false;
+            });
+
+            cfg->eraseNode(node->key);
+            if(!prev->isAdjacent(next))
```

Comment:
There is a test-case where `next` is `nullptr`. This should be caught here (for `prev`being `nullptr` too).

You can decide whether you want to throw an exception, abort the optimization or simply skip the node in such cases, depending on whether this cases breaks the optimization.